### ISSUE TITLE
csv rule file errors are printed as if they were csv file errors

### ISSUE THREAD

**Issue by t-wissmann** [simonmichael/hledger#1087]
I'm running hledger 1.14.2 from the arch linux repositories.

In csv import, if there is some error related to the csv **rule** file, then it is printed as if it were an csv syntax error in the csv data file.

# behaviour as expected

The following command works as expected:
```
$ echo '"2019/09/09","this is csv","10"' > mydata.csv
$ echo $'fields date, _, amount\ndescription MWE' > rules.txt
$ hledger -f mydata.csv --rules-file rules.txt print
2019/09/09 MWE
    unknown                     10
    income:unknown             -10

```

Let us insert a wrong csv directive:
```
$ echo 'someWrongDirective wrong' >> rules.txt
```
As expected:
```
$ hledger -f mydata.csv --rules-file rules.txt print
hledger: user error (rules.txt:3:1:
  |
3 | someWrongDirective wrong
  | ^
unexpected 's'
expecting blank or comment line, conditional block, directive, end of input, field assignment, or field name list
)

```

# Unexpected/wrong behaviour

If the csv file is not given directly, hledger thinks the error was in the csv file:
```
$ hledger -f - --rules-file rules.txt print < mydata.csv
hledger: -:1:1:
  |
1 | "2019/09/09","this is csv","10"
  | ^
unexpected '"'
expecting end of input or transaction or directive

```

And strangely this even happens when using process substitution:
```
$ hledger -f <(cat mydata.csv) --rules-file rules.txt print     
hledger: /proc/self/fd/11:1:1:
  |
1 | "2019/09/09","this is csv","10"
  | ^
unexpected '"'
expecting end of input or transaction or directive
```

**Comment by simonmichael** [comment]
This took a while to track down: with `-f -`, hledger tries all file format readers in turn until one succeeds. If the filename had a `.csv` extension, or a `csv:` prefix, it would show the csv reader’s error. With just `-`, it’s doesn’t know that it’s CSV, so just shows the error from the first reader (the journal reader). This needs some thought. Ideas welcome.

**Comment by t-wissmann** [comment]
I don't get your comment. You say that "with just `-`, it doesn't know that it's a csv". But The issue here is not in the *csv* but in the *csv rule* file which is passed via `--rules-file` and not via the `-f`.

**Comment by simonmichael** [comment]
Yes, you're right. Both kinds of error (with .csv file or with .csv.rules file) are reported by the csv reader, and hledger is ignoring the csv reader's error here.

Here's a heuristic that might work better and not be too fragile: in this situation, choose the error whose location is farthest along in the file.

**Comment by simonmichael** [comment]
While the UX was unfortunate, it's now documented ([1], [2]) that `-f -` will assume journal format and that you need `-f csv:-` for reading CSV from stdin. Attempting to guess the format was problematic in the past so I think we can consider this not a bug.

[1]: https://hledger.org/hledger.html#standard-input
[2]: https://hledger.org/hledger.html#reading-csv-from-standard-input

**Comment by t-wissmann** [comment]
Thanks! Makes perfect sense to me to specify the format along with `-`
